### PR TITLE
Jay/tssnode slash

### PR DESCRIPTION
**Pull request by byteflyfunny**

### CODE REVIEW

File: tss/node/cmd/tssnode/cmd.go [hunk 1]
```diff
@@ -12,7 +12,6 @@ func Command() *cobra.Command {
 	}
```
Comment (FletcherMan) [inline tss/node/cmd/tssnode/cmd.go:12]: command `tssnode` => `node`, as we have parent command named `tss`

File: tss/node/cmd/tssnode/cmd.go [hunk 2]
```diff
@@ -12,7 +12,6 @@ func Command() *cobra.Command {
...
 
 	cmd.AddCommand(StartCmd())
-	cmd.AddCommand(InitConfigTemplate())
 
 	cmd.PersistentFlags().StringP("config", "c", "config", "configuration file with extension")
```
Comment (FletcherMan) [inline tss/node/cmd/tssnode/cmd.go:16]: we dont need it. It is declared in parent command

File: tss/node/signer/keygen.go [hunk 3]
```diff
@@ -61,3 +75,91 @@ func (p *Processor) Keygen() {
...
+	return nil
+}
+
+// Wait for the receipt by polling the backend
+func waitForReceipt(backend *ethclient.Client, tx *etht.Transaction) (*etht.Receipt, error) {
+	t := time.NewTicker(300 * time.Millisecond)
+	receipt := new(etht.Receipt)
+	var err error
+	for range t.C {
+		receipt, err = backend.TransactionReceipt(context.Background(), tx.Hash())
```
Comment (FletcherMan) [inline tss/node/signer/keygen.go:152]: don't we need to consider the block confirmation?

File: tss/node/signer/keygen.go [hunk 4]
```diff
@@ -61,3 +75,91 @@ func (p *Processor) Keygen() {
...
+			if retries > 90 {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Wait for the receipt by polling the backend
+func waitForReceipt(backend *ethclient.Client, tx *etht.Transaction) (*etht.Receipt, error) {
```
Comment (FletcherMan) [inline tss/node/signer/keygen.go:147]: better set a timeout for the receipt waiting

File: tss/node/signer/sign_slash.go [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+					errorRes := tdtypes.NewRPCErrorResponse(req.ID, 201, "sign failed", err.Error())
+					p.wsClient.SendMsg(errorRes)
+					continue
+				}
+
+				signResponse := tsscommon.SignResponse{
+					Signature:             data,
+					SlashTxBytes:          txData,
+					SlashTxGasPrice:       gasPrice.String(),
+					SlashTxGasPriceBigInt: gasPrice,
```
Comment (FletcherMan) [inline tss/node/signer/sign_slash.go:110]: do NOT need to pass `SlashTxGasPriceBigInt`

File: tss/node/signer/sign_slash.go [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+	go func() {
+		defer func() {
+			logger.Info().Msg("exit sign process")
+		}()
+		for {
+			select {
+			case <-p.stopChan:
+				return
+			case req := <-p.signSlashChan:
+				var resId = req.ID.(tdtypes.JSONRPCStringID).String()
```
Comment (FletcherMan) [inline tss/node/signer/sign_slash.go:32]: better check with the id to prevent redundancy

File: tss/node/signer/verify.go [hunk 7]
```diff
@@ -24,24 +26,34 @@ func (p *Processor) Verify() {
...
+					logger.Error().Msg("failed to conv OffsetStartsAtIndex to int")
+					RpcResponse = tdtypes.NewRPCErrorResponse(req.ID, 201, "failed to conv OffsetStartsAtIndex to int", err.Error())
+					p.wsClient.SendMsg(RpcResponse)
+					continue
+				}
 
 				// if offset value does not equal  length of state roots array, return false
-				if len(askRequest.StateRoots) != int(askRequest.OffsetStartsAtIndex.Int64()) {
-					askResponse := tsstypes.AskResponse{
+				if len(askRequest.StateRoots) != offset {
```
Comment (FletcherMan) [inline tss/node/signer/verify.go:47]: plz remove this check. offset has no concern with the number of askRequest.StateRoots

File: tss/node/cmd/tssnode/cmd.go (hunk 1 above)
Comment (byteflyfunny) [reply tss/node/cmd/tssnode/cmd.go:35]: solved

File: tss/node/cmd/tssnode/cmd.go (hunk 2 above)
Comment (byteflyfunny) [reply tss/node/cmd/tssnode/cmd.go]: solved

File: tss/node/signer/keygen.go (hunk 3 above)
Comment (byteflyfunny) [reply tss/node/signer/keygen.go:192]: fixed

File: tss/node/signer/keygen.go (hunk 4 above)
Comment (byteflyfunny) [reply tss/node/signer/keygen.go:187]: fixed

**Pull request merged**